Bad entity attached in Join Relation

eltharin opened the pull request:
I open a new PR to be in my own branch, correction and tests are here,
report to PR https://github.com/doctrine/orm/pull/11235

greg0ire commented on tests/Tests/ORM/Hydration/ComplexeObjectHydratorTest.php:
```diff
@@ -0,0 +1,111 @@
+
+    protected function getDQL(): string
+    {
+        return 'SELECT s0_, i1_, i2_, s4_, s5_, i6_ FROM ' . Serie::class . ' s0_ LEFT JOIN s0_.itemSeries i1_ LEFT JOIN i1_.item i2_ LEFT JOIN s0_.serieImportators s4_ LEFT JOIN s4_.serie s5_ LEFT JOIN s5_.itemSeries i6_ ORDER BY i1_.serie ASC, i1_.item ASC';
```
```suggestion
        return sprintf(
            <<<'SQL'
               SELECT s0_, i1_, i2_, s4_, s5_, i6_
                 FROM %s s0_
            LEFT JOIN s0_.itemSeries i1_
            LEFT JOIN i1_.item i2_
            LEFT JOIN s0_.serieImportators s4_
            LEFT JOIN s4_.serie s5_
            LEFT JOIN s5_.itemSeries i6_
             ORDER BY i1_.serie ASC, i1_.item ASC;
            SQL
            ,
            Serie::class
        );
```

greg0ire commented on tests/Tests/ORM/Hydration/ComplexeObjectHydratorTest.php:
```diff
@@ -0,0 +1,111 @@
+            foreach ($data[$i]->getItemSeries()->toArray() as $element) {
+                self::assertEquals('Num ', substr($element->getNumber(), 0, 4));
+                self::assertEquals('Item ', substr($element->getItem()->getLibelle(), 0, 5));
+                self::assertTrue(substr($element->getNumber(), 4) === substr($element->getItem()->getLibelle(), 5));
```
Why not use `assertEquals` here as well?

eltharin replied:
habit to have a "fixed" value as expected

greg0ire commented on tests/Tests/Models/ComplexeHydration/Item.php:
```diff
@@ -0,0 +1,97 @@
+        return $this->itemSeries;
+    }
+
+    public function addItemSeries(ItemSerie $itemSeries): self
```
Please remove any unused getter, setter, adder and remover. Likewise, if a relationship doesn't need to be bidirectional for the bug to appear, please make it unidirectional

greg0ire commented on tests/Tests/Models/ComplexeHydration/ItemSerie.php:
```diff
@@ -0,0 +1,85 @@
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\Models\ComplexeHydration;
```
Should it say "complex" wherever "complexe" is used?

greg0ire commented on tests/Tests/ORM/Hydration/ComplexeObjectHydratorTest.php:
```diff
@@ -0,0 +1,111 @@
+        self::assertEquals(2, count($data));
+
+        for ($i = 0; $i < 2; $i++) {
+            foreach ($data[$i]->getItemSeries()->toArray() as $element) {
```
Maybe you could use something more concrete than `$element` here? `$itemSerie` maybe?

greg0ire commented on tests/Tests/Models/ComplexeHydration/Item.php:
```diff
@@ -0,0 +1,97 @@
+
+    /**
+     * @var Collection<int, CmsComment>
+     * @OneToMany(targetEntity="CmsComment", mappedBy="article")
```
`CmsComment`?!

greg0ire commented on tests/Tests/Models/ComplexeHydration/Item.php:
```diff
@@ -0,0 +1,97 @@
+ * @Table(name="item")
+ */
+#[Table(name: 'item')]
+#[Entity]
```
Why are there both attributes and annotations here?

greg0ire commented:
Here is a representation of the test classes for anybody trying to understand what this might be about:

```mermaid
classDiagram
    class Item {
        -string libelle
        -Collection itemSeries 
    }

    class ItemSerie {
        -string number
    }

    class Serie {
        -string libelle
    }

    class SerieImportator {
        -string libelle
    }

    Item "*" --o "1" ItemSerie

    Serie "1" o-- "*" ItemSerie
    Serie "1" o-- "*" SerieImportator
```

Personally I'm giving up for now, might get back to it later.

eltharin committed:
take good key for entity

This fixes a bug when a query has many relations
a bad entity was linked to an object instead take the good one

greg0ire commented on tests/Tests/Models/ComplexHydration/Serie.php:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     * @var Collection<int, ItemSerie>
+     * @OneToMany(targetEntity="ItemSerie", mappedBy="serie")
```
I'm asking again: why do some attributes also come with annotations?

eltharin replied:
without I have PHPCS errors : 
26 | ERROR | [ ] Property \Doctrine\Tests\Models\ComplexHydration\Serie::$itemSeries does not have @var annotation for
    |       |     its value. (SlevomatCodingStandard.TypeHints.PropertyTypeHint.MissingAnyTypeHint)
 29 | ERROR | [ ] Property \Doctrine\Tests\Models\ComplexHydration\Serie::$serieImportators does not have @var
    |       |     annotation for its value. (SlevomatCodingStandard.TypeHints.PropertyTypeHint.MissingAnyTypeHint)

eltharin replied:
oh just var is needed, I can supress OneToMany

greg0ire replied:
Yes, and there are more occurrences of more Doctrine annotations in there.

greg0ire commented on tests/Tests/ORM/Hydration/ComplexObjectHydratorTest.php:
```diff
@@ -0,0 +1,125 @@
+            LEFT JOIN s0_.itemSeries i1_
+            LEFT JOIN i1_.item i2_
+            LEFT JOIN s0_.serieImportators s4_
+            LEFT JOIN s4_.serie s5_
```
s0 is a serie, and then you join on serieImportators, and then… back on serie?!? This does not make much sense, does it? I mean, one serie has many importators, whatever that is, and importators only have one serie, so you're going to get the same object back.

greg0ire replied:
Since you reported an issue, you must have a realistic scenario that you tried to abstract away… but if it's so abstract that nobody understands the abstraction, it's not great. Tests should be simple to understand.

eltharin replied:
I agree with you, i had a automatic DQL builder and I had this but it's not useful.
But the bug apear in this case, only if there a 11+ itemSeries (Why?) and always the second item is second to last I don't really understand why

eltharin replied:
I don't know what responds, it's a bug, the correction is "just" change key() to indexof($element), 
I understand it take the last element instead of the one whitch correspond to the actual key,
It's only the tests I wrote there are difficult to understand, I just think the bug can appear in other cases so I try to repair it, now if you think it"s not helpful I can hear it,
even if I find that it remains a bug.

greg0ire replied:
You say it's a bug, but what your test shows me is that it might actually be a performance optimization built on the assumption that nobody is going to write something as stupid as `author a LEFT JOIN a.books b LEFT JOIN b.author a2`… I mean, what is the point of that?

eltharin replied:
it does not take the entity corresponding to the foreign key, Ok with a studip query but it's more a bug than a performance optimisation.
Take a look to my reproducer it's maybe more visual ; https://github.com/eltharin/errordoctrine/
the visual result : 
![303400982-0967d501-06c9-4bee-bbc7-d549f4109ea6](https://github.com/doctrine/orm/assets/7547802/cd491cd2-7c66-4236-9c17-691b532a03df)
As you can see, the last itemSeries element does'nt take item X.12 but X.2 but the foreign key is pointing to X.12
